File: sandbox/android/sampleapp/helperfuncs/src/main/java/dev/feng/helperfuncs/data/encode64.kt
package dev.feng.helperfuncs.data

import java.util.Base64


// Encode data to base64 and return the encoded string
 fun encode64(data: String): String {
     if (data.isEmpty()) {
         return ""
     }
    return Base64.getEncoder().encodeToString(data.toByteArray())
}


// Encode data to base64 and return the encoded string
fun encode64Bytes(data: ByteArray): String {
    return Base64.getEncoder().encodeToString(data)
}


// Decode base64 data to string and return a pair with the result and the decoded string, if the decoding was successful (true) or not (false)
fun decode64(data: String): Pair<Boolean, String> {
    return try {
        val decodedBytes = Base64.getDecoder().decode(data)
        Pair(true, String(decodedBytes))
    } catch (e: Exception) {
        Pair(false, "")
    }
}

// Decode base64 data to bytes array and return a pair with the result and the decoded bytes array, if the decoding was successful (true) or not (false)
fun decode64Bytes(data: String): Pair<Boolean, ByteArray> {
    return try {
        val decodedBytes = Base64.getDecoder().decode(data)
        Pair(true, decodedBytes)
    } catch (e: Exception) {
        Pair(false, ByteArray(0))
    }
}

File: data/encode64.go
package data

import "encoding/base64"

// Encode64Bytes returns the base64 encoding of the given byte slice.
func Encode64Bytes(data []byte) string {
	return base64.StdEncoding.EncodeToString(data)
}

// Encode64 returns the base64 encoding of the given string.
func Encode64(data string) string {
	return Encode64Bytes([]byte(data))
}

// Decode64Bytes decodes the base64 encoded string and returns the corresponding byte slice.
func Decode64Bytes(data string) ([]byte, error) {
	return base64.StdEncoding.DecodeString(data)
}

// Decode64 decodes the base64 encoded string and returns the corresponding decoded string.
func Decode64(data string) (string, error) {
	decodedBytes, err := Decode64Bytes(data)
	if err != nil {
		return "", err
	}
	return string(decodedBytes), nil
}
